spl/src/token_interface.rs

```diff
@@ -0,0 +1,71 @@
+use anchor_lang::solana_program::pubkey::Pubkey;
+use std::ops::Deref;
+
+static IDS: [Pubkey; 2] = [spl_token::ID, spl_token_2022::ID];
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct TokenAccount(spl_token_2022::state::Account);
+
+impl anchor_lang::AccountDeserialize for TokenAccount {
+    fn try_deserialize_unchecked(buf: &mut &[u8]) -> anchor_lang::Result<Self> {
+        spl_token_2022::extension::StateWithExtensions::<spl_token_2022::state::Account>::unpack(
+            buf,
+        )
+        .map(|t| TokenAccount(t.base))
+        .map_err(Into::into)
+    }
+}
+
+impl anchor_lang::AccountSerialize for TokenAccount {}
+
+impl anchor_lang::Owners for TokenAccount {
+    fn owners() -> &'static [Pubkey] {
+        &IDS
+    }
+}
+
+impl Deref for TokenAccount {
+    type Target = spl_token_2022::state::Account;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Mint(spl_token_2022::state::Mint);
+
+impl anchor_lang::AccountDeserialize for Mint {
+    fn try_deserialize_unchecked(buf: &mut &[u8]) -> anchor_lang::Result<Self> {
+        spl_token_2022::extension::StateWithExtensions::<spl_token_2022::state::Mint>::unpack(buf)
+            .map(|t| Mint(t.base))
+            .map_err(Into::into)
+    }
+}
+
+impl anchor_lang::AccountSerialize for Mint {}
+
+impl anchor_lang::Owners for Mint {
+    fn owners() -> &'static [Pubkey] {
+        &IDS
+    }
+}
+
+impl Deref for Mint {
+    type Target = spl_token_2022::state::Mint;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[derive(Clone)]
+pub struct TokenInterface;
```

Comment:
More of a general usage question / thinking outloud. It's up to the dev to implement checks that any `InterfaceAccount` and `Interface` both refer to the same program? And in general the `Interface<T>` types and the `InterfaceAccount<T>` generic types don't really know about each other? Like `token_2022::TokenAccount` doesn't know that it's an interface account for the `TokenInterface` type.

Reply:
@Henry-E - I am preparing https://github.com/joncinque/anchor/pull/1/files separately for specifying the token_program in this case. 

This is a special case since the `token_program` field is required for `TokenAccount` init.

Using the interfaces myself - I think an `owner` check or a straight `has_one` against some state will cover most other cases.

Reply:
> More of a general usage question / thinking outloud. It's up to the dev to implement checks that any InterfaceAccount and Interface both refer to the same program?

This is a really great point. We could require that `InterfaceAccount` has an owner check, unless that would be too annoying. I don't feel too strongly either way, but leaning towards leaving it as is and letting people add explicit `owner` checks.

Reply:
We chatted about this on telegram, the `token::token_program = ` constraint by elliot will be designed to allow different program accounts to be specified as the `token_program` account used. This is necessary because there could be multiple types of accounts working.

One edge case that comes to mind is does using `token_program: Interface<'info, TokenInterface>` still work with the existing `token::` constraints for initializing and checking the mint of a token account. I have to check if any of the tests do this, it seems likely they would.

It looks like this subPR hasn't been merged into the main PR we're looking at now, so maybe it's best to just propose it as a new and separate PR directly to anchor once the interface PR has been merged.

Reply:
Ok, I found the `ProxyCreateTokenAccount` account context in the escrow test and it uses the `token::` constraints to init an account with an interface as the token_program.

Reply:
This PR is already a bit big for my taste, so I'm happy to look at the `token_program` constraint separately, and I'll be very happy to give a review too